packages/core-plugin-api/src/plugin/Plugin.tsx

```diff
@@ -58,6 +67,16 @@ export class PluginImpl<
     return extension.expose(this);
   }
 
+  reconfigure(fn: ReconfigureFunction): void {
```

Comment:
Bit flipped around here right? With this the owner of the options transform will be the one reconfiguring the plugin, but that's something we want the plugin to own. I'm thinking these kinda building blocks:

```ts
// integration code
const myNewPlugin = somePlugin.reconfigure({ deltaX: 1 });

// plugin code
interface PluginInputOptions { deltaX: number }
interface PluginOptions { x: number }

const somePlugin = createPlugin({
  id: 'some',
  configure(options?: PluginInputOptions): PluginOptions {
    if (!options) {
      return DEFAULT_OPTIONS
    }
    return { x: DEFAULT_OPTIONS.x + options.deltaX }
  }
})
```

The `configure` method would be responsible for generating the default config, or if `reconfigure` has been called, then how to merge those options together with the default ones. In the end it is the return type of `configure` that ends up being used as the actual options within the plugin.

It's something along these lines that I'm thinking anyway, but it's possible it doesn't quite work out or that there are better options.

Reply:
The `PluginInputOptions` would be what an integrator passes to `reconfigure`, they then get routed to the `configure` method by some logic within the `createPlugin` method/plugin instance.

In my example `PluginInputOptions` isn't compatible with `DEFAULT_OPTIONS` right? The default options are of the type `PluginOptions`. `PluginOptions` is what the plugin consumes internally, while `PluginInputOptions` are what people provide from the outside. They may be the same type, but they may also be different. That split makes some use-cases much cleaner, like merging with or mutating default options through the input options. It's also hugely important for API stability and keeping sanity within the plugin itself, as all backwards compatibility code can be handled within the `configure` method, as opposed to being spread out throughout the plugin.

Reply:
Ah ok, got it. I was looking at it a bit differently but of course can be done this way too.

I see that you guys working on https://github.com/backstage/backstage/pull/11819, does this PR still make sense?

Reply:
#11819 solves a slightly different problem imo, although it could still be that these intersect a bit. I think it makes sense to move both PRs forward either way, figuring out how either fits in.

Reply:
@Rugvip 

Started doing it, and it doesn't add up to one whole thing. 

With the provided example of:
```
// integration code
const myNewPlugin = somePlugin.reconfigure({ deltaX: 1 });

// plugin code
interface PluginInputOptions { deltaX: number }
interface PluginOptions { x: number }

const somePlugin = createPlugin({
  id: 'some',
  configure(options?: PluginInputOptions): PluginOptions {
    if (!options) {
      return DEFAULT_OPTIONS
    }
    return { x: DEFAULT_OPTIONS.x + options.deltaX }
  }
})
```

1. From the explanation above `configure` will be called with `PluginInputOptions` only from `reconfigure`.
2. `configure` on the example has already defined strategy how to merge, but we would like to allow the integrator of the plugin to merge options as he/she wants.

So it means that `configure` function we don't want to have and `reconfigure` should get as a parameter a function, not an object.

I would still stick with the current implementation. 
The only change what we could do is:

```
export type AnyPluginInputOptions = { [name: string]: unknown };

export type ReconfigureFunction = (
  options: AnyPluginInputOptions,
) => AnyPluginOptions;
```

So it means that what default options are defined in the default options would be of different type that `reconfigure` can apply on top of it.

Reply:
The idea is that the configure function would be defined by the plugin author and it is not up for customization by integrators. The only thing that is available to integrators would be the input options